dali/kernels/shape.h

```diff
@@ -0,0 +1,430 @@
+
+template <>
+struct TensorShape<DynamicDimensions> {
+  // TODO overkill like standard containers with all the iterator, elemet_types, reference_type etc?
+  // or just use decltype(shape.begin()) begin() { return shape.begin(); } ?
+  using element_type = int64_t;
+  using value_type = int64_t;
+  using size_type = int;
+  using reference = value_type &;
+  using const_reference = const value_type &;
+  using container_type = std::vector<element_type>;
+  using iterator = typename container_type::iterator;
+  using const_iterator = typename container_type::const_iterator;
+  using reverse_iterator = typename container_type::reverse_iterator;
+  using const_reverse_iterator = typename container_type::const_reverse_iterator;
+
+  TensorShape(const std::vector<int64_t> &s) : shape(s) {}
+
+  template <size_t N>
+  TensorShape(std::array<int64_t, N> s) : shape(s.begin(), s.end()) {}
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {}
+
+  TensorShape() = default;
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <int other_ndim>
+  TensorShape(const TensorShape<other_ndim> &other)
+      : shape(other.shape.begin(), other.shape.end()) {}
+
+  template <int other_ndim>
+  TensorShape &operator=(const TensorShape<other_ndim> &other) {
+    shape = std::vector<int64_t>(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  // We allow only explicit conversion from dynamic to static dim
+  template <int other_ndim>
+  TensorShape<other_ndim> to_static_ndim() const {
+    static_assert(other_ndim != DynamicDimensions,
+                  "Conversion to static only allowed for static shape");
+    assert(size() == other_ndim);
+    TensorShape<other_ndim> shape;
+    for (int i = 0; i < other_ndim; i++) {
+      shape[i] = (*this)[i];
+    }
+    return shape;
+  }
+
+  reference operator[](int d) { return shape[d]; }
+  const_reference operator[](int d) const { return shape[d]; }
+
+  iterator begin() noexcept { return shape.begin(); }
+  iterator end() noexcept { return shape.end(); }
+  const_iterator begin() const noexcept { return shape.begin(); }
+  const_iterator end() const noexcept { return shape.end(); }
+  const_iterator cbegin() const noexcept { return shape.cbegin(); }
+  const_iterator cend() const noexcept { return shape.cend(); }
+  reverse_iterator rbegin() noexcept { return shape.rbegin(); }
+  reverse_iterator rend() noexcept { return shape.rend(); }
+  const_reverse_iterator rbegin() const noexcept { return shape.rbegin(); }
+  const_reverse_iterator rend() const noexcept { return shape.rend(); }
+  const_reverse_iterator crbegin() const noexcept { return shape.crbegin(); }
+  const_reverse_iterator crend() const noexcept { return shape.crend(); }
+
+  TensorShape<DynamicDimensions> first(int count) {
```

Comment:
first/last -> outer/inner?

Reply:
OK, I won't argue. Neither is particularly self-explanatory, so maybe a comment would be nice.
What would be quite welcome is to have a statically-sized variant of these.